diff -ruN binutils/bfd/configure binutils.bak/bfd/configure
--- binutils/bfd/configure	2012-09-23 01:02:58.000000000 -0500
+++ binutils.bak/bfd/configure	2015-06-20 22:30:34.088466082 -0500
@@ -10703,21 +10703,9 @@
 
   # Need _noncanonical variables for this.
 
-
-
-
-  # libbfd.a is a host library containing target dependent code
-  bfdlibdir='$(libdir)'
-  bfdincludedir='$(includedir)'
-  if test "${host}" != "${target}"; then
-    bfdlibdir='$(exec_prefix)/$(host_noncanonical)/$(target_noncanonical)/lib'
-    bfdincludedir='$(exec_prefix)/$(host_noncanonical)/$(target_noncanonical)/include'
-  fi
-
-
-
-
-
+# libbfd.a is a host library containing target dependent code
+bfdlibdir='$(exec_prefix)/host/lib'
+bfdincludedir='$(exec_prefix)/host/include'
 
 host64=false
 target64=false
diff -ruN binutils/opcodes/configure binutils.bak/opcodes/configure
--- binutils/opcodes/configure	2012-09-23 01:03:01.000000000 -0500
+++ binutils.bak/opcodes/configure	2015-06-20 22:31:22.316466235 -0500
@@ -9847,23 +9847,11 @@
   INSTALL_LIBBFD_FALSE=
 fi
 
-  # Need _noncanonical variables for this.
-
-
-
-
-  # libbfd.a is a host library containing target dependent code
-  bfdlibdir='$(libdir)'
-  bfdincludedir='$(includedir)'
-  if test "${host}" != "${target}"; then
-    bfdlibdir='$(exec_prefix)/$(host_noncanonical)/$(target_noncanonical)/lib'
-    bfdincludedir='$(exec_prefix)/$(host_noncanonical)/$(target_noncanonical)/include'
-  fi
-
-
-
-
+# Need _noncanonical variables for this.
 
+# libbfd.a is a host library containing target dependent code
+bfdlibdir='$(exec_prefix)/host/lib'
+bfdincludedir='$(exec_prefix)/host/include'
 
 # host-specific stuff:
 
