diff -dur a/gcc/config.gcc b/gcc/config.gcc
--- a/gcc/config.gcc	2007-02-01 21:57:17.000000000 +0100
+++ b/gcc/config.gcc	2007-02-01 22:11:06.000000000 +0100
@@ -690,7 +690,7 @@
 	    default_use_cxa_atexit=yes
 	    ;;
 	*)
-	    tmake_file="$tmake_file arm/t-linux"
+	    tmake_file="$tmake_file arm/t-linux arm/t-arm-elf"
 	    ;;
 	esac
 	tm_file="$tm_file arm/aout.h arm/arm.h"
diff -dur a/gcc/config/arm/linux-elf.h b/gcc/config/arm/linux-elf.h
--- a/gcc/config/arm/linux-elf.h	2007-02-01 21:57:17.000000000 +0100
+++ b/gcc/config/arm/linux-elf.h	2007-02-01 23:00:42.000000000 +0100
@@ -63,7 +63,7 @@
    %{shared:-lc} \
    %{!shared:%{profile:-lc_p}%{!profile:-lc}}"
 
-#define LIBGCC_SPEC "%{msoft-float:-lfloat} %{mfloat-abi=soft*:-lfloat} -lgcc"
+#define LIBGCC_SPEC "-lgcc"
 
 #ifdef USE_UCLIBC
 #define LINUX_TARGET_INTERPRETER "/lib/ld-uClibc.so.0"
