diff -ru gcc.bak/gcc/config.gcc gcc-4.1.2/gcc/config.gcc
--- gcc.bak/gcc/config.gcc	2006-10-15 19:12:23.000000000 -0400
+++ gcc-4.1.2/gcc/config.gcc	2007-07-04 21:07:42.000000000 -0400
@@ -685,7 +685,7 @@
 	    default_use_cxa_atexit=yes
 	    ;;
 	*)
-	    tmake_file="$tmake_file arm/t-linux"
+	    tmake_file="$tmake_file arm/t-linux arm/t-arm-elf"
 	    ;;
 	esac
 	tm_file="$tm_file arm/aout.h arm/arm.h"
diff -ru gcc.bak/gcc/config/arm/linux-elf.h gcc-4.1.2/gcc/config/arm/linux-elf.h
--- gcc.bak/gcc/config/arm/linux-elf.h	2005-10-09 21:04:31.000000000 -0400
+++ gcc-4.1.2/gcc/config/arm/linux-elf.h	2007-07-04 21:07:42.000000000 -0400
@@ -49,7 +49,7 @@
    %{shared:-lc} \
    %{!shared:%{profile:-lc_p}%{!profile:-lc}}"
 
-#define LIBGCC_SPEC "%{msoft-float:-lfloat} %{mfloat-abi=soft*:-lfloat} -lgcc"
+#define LIBGCC_SPEC "-lgcc"
 
 #define GLIBC_DYNAMIC_LINKER "/lib/ld-linux.so.2"
 
