--- linux-old/init/main.c	2005-09-09 21:42:58.000000000 -0500
+++ linux-new/init/main.c	2005-10-24 02:07:37.683498720 -0500
@@ -242,15 +242,6 @@
 	if (obsolete_checksetup(param))
 		return 0;
 
-	/*
-	 * Preemptive maintenance for "why didn't my mispelled command
-	 * line work?"
-	 */
-	if (strchr(param, '.') && (!val || strchr(param, '.') < val)) {
-		printk(KERN_ERR "Unknown boot option `%s': ignoring\n", param);
-		return 0;
-	}
-
 	if (panic_later)
 		return 0;
 
