diff -ru linux/drivers/serial/serial_core.c linux2/drivers/serial/serial_core.c
--- linux/drivers/serial/serial_core.c	2009-12-02 21:51:21.000000000 -0600
+++ linux2/drivers/serial/serial_core.c	2009-12-08 06:17:06.000000000 -0600
@@ -113,7 +113,7 @@
 static void uart_tasklet_action(unsigned long data)
 {
 	struct uart_state *state = (struct uart_state *)data;
-	tty_wakeup(state->port.tty);
+	if (state->port.tty) tty_wakeup(state->port.tty);
 }
 
 static inline void
