--- linux-orig/init/main.c	2007-10-09 15:31:38.000000000 -0500
+++ linux-2.6.23/init/main.c	2007-11-05 00:56:14.000000000 -0600
@@ -286,15 +286,6 @@
 	if (obsolete_checksetup(param))
 		return 0;
 
-	/*
-	 * Preemptive maintenance for "why didn't my misspelled command
-	 * line work?"
-	 */
-	if (strchr(param, '.') && (!val || strchr(param, '.') < val)) {
-		printk(KERN_ERR "Unknown boot option `%s': ignoring\n", param);
-		return 0;
-	}
-
 	if (panic_later)
 		return 0;
 
