--- a/linux/arm/syscallent.h
+++ b/linux/arm/syscallent.h
@@ -431,7 +431,7 @@
 	{ 5,	0,	printargs,		"SYS_398"	}, /* 398 */
 	{ 5,	0,	printargs,		"SYS_399"	}, /* 399 */
 
-#if SYS_socket_subcall != 400
+#if defined(SYS_socket_subcall) && (SYS_socket_subcall != 400)
  #error fix me
 #endif
 	{ 8,	0,	printargs,		"socket_subcall"}, /* 400 */
@@ -453,7 +453,7 @@
 	{ 5,	TN,	sys_sendmsg,		"sendmsg"	}, /* 416 */
 	{ 5,	TN,	sys_recvmsg,		"recvmsg"	}, /* 417 */
 
-#if SYS_ipc_subcall != 418
+#if defined(SYS_ipc_subcall) && (SYS_ipc_subcall != 418)
  #error fix me
 #endif
 	{ 4,	0,	printargs,		"ipc_subcall"	}, /* 418 */
