Ugly workaround until I can switch to musl.

diff -r 9e105bab92e5 scripts/genconfig.sh
--- a/scripts/genconfig.sh	Wed Apr 16 08:54:19 2014 -0500
+++ b/scripts/genconfig.sh	Wed Apr 16 14:53:56 2014 -0500
@@ -34,6 +34,11 @@
     #error nope
     #endif
 EOF
+
+  # Hard to come by in uClibc.
+  probesymbol TOYBOX_ICONV -c << EOF
+    #include "iconv.h"
+EOF
 }
 
 genconfig()
diff -r 9e105bab92e5 toys/pending/iconv.c
--- a/toys/pending/iconv.c	Wed Apr 16 08:54:19 2014 -0500
+++ b/toys/pending/iconv.c	Wed Apr 16 14:53:56 2014 -0500
@@ -9,6 +9,7 @@
 config ICONV
   bool "iconv"
   default y
+  depends on TOYBOX_ICONV
   help
     usage: iconv [-f FROM] [-t TO] [FILE...]
 
