diff -ruN uClibc/libc/sysdeps/linux/common/bits/uClibc_locale.h uClibc.bak/libc/sysdeps/linux/common/bits/uClibc_locale.h
--- uClibc/libc/sysdeps/linux/common/bits/uClibc_locale.h	2010-04-02 10:34:27.000000000 -0500
+++ uClibc.bak/libc/sysdeps/linux/common/bits/uClibc_locale.h	2010-11-10 15:11:31.000000000 -0600
@@ -41,11 +41,9 @@
 
 # define __LOCALE_C_ONLY
 
-# ifdef _LIBC
 #  define __XL_NPP(N) N
 #  define __LOCALE_PARAM
 #  define __LOCALE_ARG
-# endif
 
 #endif
 
@@ -90,8 +88,6 @@
 struct __uclibc_locale_struct;
 typedef struct __uclibc_locale_struct *__locale_t;
 
-#ifdef _LIBC
-
 /* extern void _locale_set(const unsigned char *p); */
 /* extern void _locale_init(void); */
 
@@ -367,8 +363,6 @@
 #endif
 /**********************************************************************/
 
-#endif /* _LIBC */
-
 #endif /* !defined(__LOCALE_C_ONLY) */
 
 #endif /* _UCLIBC_LOCALE_H */
