--- a/libc/sysdeps/linux/arm/Makefile.arch	2007-03-06 02:06:26.000000000 -0600
+++ b/libc/sysdeps/linux/arm/Makefile.arch	2009-04-06 02:59:26.000000000 -0500
@@ -15,7 +15,7 @@
 ifeq ($(CONFIG_ARM_EABI),y)
 CSRC += aeabi_assert.c aeabi_atexit.c aeabi_errno_addr.c \
 	aeabi_localeconv.c aeabi_memclr.c aeabi_memcpy.c \
-	aeabi_memmove.c aeabi_memset.c find_exidx.c
+	aeabi_memmove.c aeabi_memset.c find_exidx.c aeabi_unwind_cpp_pr1.c
 SSRC += syscall-eabi.S
 ARCH_OBJ_FILTEROUT := syscall.c
 ifeq ($(UCLIBC_HAS_WCHAR),y)
