diff -ru uClibc/libc/sysdeps/linux/sparc/bits/kernel_types.h uClibc.bak/libc/sysdeps/linux/sparc/bits/kernel_types.h
--- uClibc/libc/sysdeps/linux/sparc/bits/kernel_types.h	2008-07-23 06:23:36.000000000 -0500
+++ uClibc.bak/libc/sysdeps/linux/sparc/bits/kernel_types.h	2009-03-26 18:05:26.000000000 -0500
@@ -4,9 +4,10 @@
  * our private content, and not the kernel header, will win.
  *  -Erik
  */
-#if ! defined __ARCH_SPARC_POSIX_TYPES_H && ! defined __ARCH_SPARC64_POSIX_TYPES_H
+#if ! defined __ARCH_SPARC_POSIX_TYPES_H && ! defined __ARCH_SPARC64_POSIX_TYPES_H && !defined __SPARC_POSIX_TYPES_H
 #define __ARCH_SPARC_POSIX_TYPES_H
 #define __ARCH_SPARC64_POSIX_TYPES_H
+#define __SPARC_POSIX_TYPES_H
 
 # if __WORDSIZE == 64
 typedef unsigned long          __kernel_size_t;
