diff -ru uClibc/libc/sysdeps/linux/sh/bits/kernel_types.h uClibc2/libc/sysdeps/linux/sh/bits/kernel_types.h
--- uClibc/libc/sysdeps/linux/sh/bits/kernel_types.h	2008-07-23 06:23:36.000000000 -0500
+++ uClibc2/libc/sysdeps/linux/sh/bits/kernel_types.h	2009-09-17 21:05:29.000000000 -0500
@@ -4,8 +4,9 @@
  * our private content, and not the kernel header, will win.
  *  -Erik
  */
-#ifndef __ASM_SH_POSIX_TYPES_H
+#if !defined(__ASM_SH_POSIX_TYPES_H) && !defined(__ASM_SH_POSIX_TYPES_32_H)
 #define __ASM_SH_POSIX_TYPES_H
+#define __ASM_SH_POSIX_TYPES_32_H
 
 typedef unsigned short	__kernel_dev_t;
 typedef unsigned long	__kernel_ino_t;
diff -ru uClibc/libc/sysdeps/linux/sh64/bits/kernel_types.h uClibc2/libc/sysdeps/linux/sh64/bits/kernel_types.h
--- uClibc/libc/sysdeps/linux/sh64/bits/kernel_types.h	2008-07-23 06:23:36.000000000 -0500
+++ uClibc2/libc/sysdeps/linux/sh64/bits/kernel_types.h	2009-09-17 21:06:35.000000000 -0500
@@ -16,8 +16,9 @@
  * our private content, and not the kernel header, will win.
  *  -Erik
  */
-#ifndef __ASM_SH64_POSIX_TYPES_H
+#if !defined(__ASM_SH64_POSIX_TYPES_H) && !defined(__ASM_SH_POSIX_TYPES_64_H)
 #define __ASM_SH64_POSIX_TYPES_H
+#define __ASM_SH_POSIX_TYPES_64_H
 
 typedef unsigned short  __kernel_dev_t;
 typedef unsigned long   __kernel_ino_t;
