diff -r 34a75fcd5a9b Makefile.in
--- a/Makefile.in	Sun Mar 11 23:06:39 2007 +0000
+++ b/Makefile.in	Fri Mar 16 15:47:35 2007 -0400
@@ -36,7 +36,7 @@ include $(top_srcdir)extra/locale/Makefi
 # last included to catch all the objects added by others (locales/threads)
 include $(top_srcdir)libc/Makefile.in
 
-include/bits/uClibc_config.h: extra/config/conf .config
+include/bits/uClibc_config.h: kconfig/conf .config
 	$(Q)$(INSTALL) -d $(dir $@)
 	$(Q)@$< -o $(top_srcdir)extra/Configs/Config.in
 	$(top_srcdir)extra/scripts/conf-header.sh .config > include/bits/uClibc_config.h
@@ -290,42 +291,7 @@ include/bits:
 include/bits:
 	$(INSTALL) -d include/bits
 
-# configuration
-# ---------------------------------------------------------------------------
-extra/config/conf extra/config/mconf: include/bits
-	$(Q)$(MAKE) -C extra/config $(notdir $@)
-
-menuconfig: extra/config/mconf include/bits
-	$(Q)./extra/config/mconf extra/Configs/Config.in
-
-config: extra/config/conf include/bits
-	$(Q)./extra/config/conf extra/Configs/Config.in
-
-oldconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -o extra/Configs/Config.in
-
-silentoldconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -s extra/Configs/Config.in
-
-randconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -r extra/Configs/Config.in
-
-allyesconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -y extra/Configs/Config.in
-	sed -i -e "s/^DODEBUG=.*/# DODEBUG is not set/" .config
-	sed -i -e "s/^DOASSERTS=.*/# DOASSERTS is not set/" .config
-	sed -i -e "s/^SUPPORT_LD_DEBUG_EARLY=.*/# SUPPORT_LD_DEBUG_EARLY is not set/" .config
-	sed -i -e "s/^SUPPORT_LD_DEBUG=.*/# SUPPORT_LD_DEBUG is not set/" .config
-	sed -i -e "s/^UCLIBC_MJN3_ONLY=.*/# UCLIBC_MJN3_ONLY is not set/" .config
-	$(Q)./extra/config/conf -o extra/Configs/Config.in
-
-allnoconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -n extra/Configs/Config.in
-
-defconfig: extra/config/conf include/bits
-	$(Q)./extra/config/conf -d extra/Configs/Config.in
-
-clean:
+clean::
 	$(Q)$(RM) -r lib include/bits
 	$(RM) lib*/*.a ldso/*/*.a libpthread/*/*.a
 	$(RM) include/fpu_control.h include/dl-osinfo.h include/hp-timing.h
@@ -349,7 +315,6 @@ distclean: clean
 	-find . \( -name core -o -name \*.orig -o -name \*~ \) -exec $(RM) {} \;
 	$(RM) .config .config.old .config.cmd
 	$(RM) extra/locale/*.tgz
-	$(MAKE) -C extra/config distclean
 
 dist release:
 	$(RM) -r ../uClibc-$(VERSION) ../uClibc-$(VERSION).tar.bz2
@@ -359,3 +324,5 @@ dist release:
 
 test check:
 	$(Q)$(MAKE) -C test
+
+include $(top_srcdir)kconfig/Makefile
diff -r 34a75fcd5a9b Makerules
--- a/Makerules	Sun Mar 11 23:06:39 2007 +0000
+++ b/Makerules	Mon Mar 12 15:35:07 2007 -0400
@@ -279,7 +279,7 @@ CRTS_COMPAT :=
 	$(do_ar)
 
 .PHONY: dummy create
-clean: objclean-y headers_clean-y
+clean:: objclean-y headers_clean-y
 
 objclean-y: $(objclean-y)
 headers_clean-y: $(headers_clean-y)
diff -r 34a75fcd5a9b extra/Makefile.in
--- a/extra/Makefile.in	Sun Mar 11 23:06:39 2007 +0000
+++ b/extra/Makefile.in	Mon Mar 12 15:47:38 2007 -0400
@@ -6,4 +6,3 @@
 #
 
 include $(top_srcdir)extra/locale/Makefile.in
-include $(top_srcdir)extra/config/Makefile.in
